File: examples/python/input/shortcuts_theme.py
#!/usr/bin/env python3
"""Custom keyboard shortcuts on a widget and the process-wide input theme.

    make run-example EX=python/input/shortcuts_theme

Keys: F2 archives (closes, reported via fired()), Ctrl-X deletes the
highlighted item in place, Enter picks.
"""

from __future__ import annotations

import sparcli as sc

ACTION_ARCHIVE = 1
NOTES = ['Meeting notes', 'Shopping list', 'Project ideas', 'Travel plans']


def apply_theme() -> None:
    """Process-wide defaults inherited by every widget for unset options."""
    sc.set_theme(sc.Theme(accent=sc.Color.MAGENTA,
                          prompt_style=sc.Style.bold(sc.Color.MAGENTA),
                          cursor_marker='▶ ', marker='  ',
                          hint_layout=sc.HintLayout.INLINE))


def run_select_with_shortcuts() -> None:
    """A select with a RETURN shortcut (F2) and a CALLBACK shortcut (Ctrl-X)."""
    select = sc.Select(sc.SelectOpts(prompt='Pick a note'))
    for note in NOTES:
        select.add(note)

    def delete_highlighted() -> bool:
        # Runs in place; returning True (or None) keeps the prompt open.
        select.remove(select.cursor())
        return True

    shortcuts = (sc.Shortcuts()
                 .on_return(sc.key_fn(2), ACTION_ARCHIVE, name='archive')
                 .on_callback(sc.key_ctrl('x'), delete_highlighted,
                              name='delete'))

    # Re-create the select with the shortcuts wired in (opts are read at run).
    select = sc.Select(sc.SelectOpts(prompt='Pick a note',
                                     shortcuts=shortcuts))
    for note in NOTES:
        select.add(note)

    index = select.run_one()
    if index is None:
        return
    label = select.label(index)
    if shortcuts.fired() == ACTION_ARCHIVE:
        sc.println(f'  -> archive {label!r}')
    else:
        sc.println(f'  -> picked {label!r}')


def main() -> None:
    if not sc.input_available():
        sc.alert.warning('Run this example in a real terminal (not piped).')
        return
    apply_theme()
    try:
        run_select_with_shortcuts()
    finally:
        sc.set_theme(None)  # restore the built-in defaults


if __name__ == '__main__':
    main()

File: examples/c/input/shortcuts_theme.c
/*
 * shortcuts_theme.c - custom keyboard shortcuts on a widget and the
 * process-wide input theme.
 *
 * Run (after `make`) in a real terminal:
 *   make run-example EX=c/input/shortcuts_theme
 *
 * Keys in the select: F2 archives (closes and reports), Ctrl-X deletes the
 * highlighted item in place, Enter picks, Esc cancels.
 */

#include <sparcli.h>

#include <stdio.h>


/** Shortcut id reported through out_shortcut_id. */
enum { ACTION_ARCHIVE = 1 };


static void apply_theme(void);
static void run_select_with_shortcuts(void);
static bool delete_highlighted(int id, void *user);


int main(void) {
    if (!sc_input_available()) {
        sc_alert_warning("Run this example in a real terminal "
                         "(not under a pipe).");
        return 0;
    }

    apply_theme();
    run_select_with_shortcuts();

    // Reset to the built-in defaults for any code that runs afterwards.
    sc_input_set_theme(NULL);
    return 0;
}

/** Process-wide defaults: every widget inherits them for zero-init opts. */
static void apply_theme(void) {
    sc_input_set_theme(&(ScInputTheme){
        .accent        = SC_ANSI_COLOR_MAGENTA,
        .prompt_style  = { SC_TEXT_ATTR_BOLD, SC_ANSI_COLOR_MAGENTA,
                           SC_ANSI_COLOR_NONE },
        .cursor_marker = "▶ ",
        .marker        = "  ",
        .hint_layout   = SC_HINT_INLINE,
    });
}

/** CALLBACK shortcut: remove the highlighted item, keep the prompt open. */
static bool delete_highlighted(int id, void *user) {
    (void)id;
    ScSelect *select = user;
    sc_select_remove(select, sc_select_cursor(select));
    return true;    // true = stay open
}

/** A select with one RETURN shortcut and one CALLBACK shortcut. */
static void run_select_with_shortcuts(void) {
    int fired = -1;
    ScShortcut shortcuts[] = {
        // RETURN: closes the prompt and reports the id via out_shortcut_id.
        { .chord      = sc_key_fn(2),
          .id         = ACTION_ARCHIVE,
          .mode       = SC_SHORTCUT_RETURN,
          .hint_label = "archive" },
        // CALLBACK: runs in place while the prompt stays open.
        { .chord      = sc_key_ctrl('x'),
          .mode       = SC_SHORTCUT_CALLBACK,
          .on_fire    = delete_highlighted,
          .hint_label = "delete" },
    };

    ScSelect *select = sc_select_new((ScSelectOpts){
        .prompt          = "Pick a note",
        .shortcuts       = shortcuts,
        .n_shortcuts     = 2,
        .out_shortcut_id = &fired,
    });
    // The delete callback needs the live handle.
    shortcuts[1].user = select;

    static const char *const NOTES[] = {
        "Meeting notes", "Shopping list", "Project ideas", "Travel plans",
    };
    for (size_t i = 0; i < sizeof NOTES / sizeof NOTES[0]; i++) {
        sc_select_add(select, NOTES[i]);
    }

    size_t picked = 0;
    size_t count  = 1;
    ScInputStatus status = sc_select_run(select, &picked, &count);

    if (status == SC_INPUT_OK && fired == ACTION_ARCHIVE && count > 0) {
        printf("  -> archive \"%s\"\n", sc_select_label(select, picked));
    } else if (status == SC_INPUT_OK && count > 0) {
        printf("  -> picked \"%s\"\n", sc_select_label(select, picked));
    } else if (status == SC_INPUT_OK) {
        printf("  -> nothing left to pick\n");
    }

    sc_select_free(select);
}
